### PR TITLE
agent: Log ttrpc messages

### PR DESCRIPTION
**Pull request by jodh-intel**
The `ttrpc` crate uses the `log` crate for logging. But the agent uses the `slog` crate. This means that currently, all `ttrpc` log messages are being discarded.

Use the `slog-stdlog` create to redirect `log` crate logging calls into `slog` so they are visible in the agents log output.

Also adds a new `KATA_AGENT_LOG_LEVEL` environment variable to make testing this easier, plus a bunch of new tests.

Fixes: #978.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

File: src/agent/src/config.rs [hunk 1]
```diff
@@ -22,6 +22,7 @@ const DEFAULT_CONTAINER_PIPE_SIZE: i32 = 0;
 const VSOCK_ADDR: &str = "vsock://-1";
 const VSOCK_PORT: u16 = 1024;
 const SERVER_ADDR_ENV_VAR: &str = "KATA_AGENT_SERVER_ADDR";
+const LOG_LEVEL_ENV_VAR: &str = "KATA_LOG_LEVEL";
```
Comment (fidencio) [inline src/agent/src/config.rs:25]: @jodh-intel, firstly, thanks for the patches!

One question, would make sense to just add a comment explicitly mentioning that the env-var shoud only be used for testing purposes?

**Review by fidencio — APPROVED** [review]
lgtm!

File: src/agent/src/config.rs (hunk 1 above)
Comment (devimc) [reply src/agent/src/config.rs:25]: also, I wonder if this variable should start with `KATA_AGENT_`

Comment (jodh-intel) [reply src/agent/src/config.rs:27]: Sure - done.

Comment (jodh-intel) [reply src/agent/src/config.rs]: @devimc - I did think about that but decided against it. But now you've convinced me it should have an agent prefix - branch updated ;)

**Review by devimc — APPROVED** [review]
lgtm - thanks @jodh-intel

**Comment by jodh-intel** [comment]
Note that with this PR applied, the agent produces quite a few messages like this, so we may want to change the log level of these in the `ttrpc` crate:

```
{"msg":"Failed to try send \"Full(..)\"","level":"WARN","ts":"2020-10-16T15:55:37.638214410+01:00","name":"kata-agent","subsystem":"rpc","source":"agent","pid":"4732","version":"0.1.0"}
{"msg":"client thread quit","level":"INFO","ts":"2020-10-16T15:55:37.639078762+01:00","source":"agent","subsystem":"rpc","version":"0.1.0","name":"kata-agent","pid":"4732"}
```

/cc @lifupan

**Comment by jodh-intel** [comment]
@lifupan - it's a slight abuse of logging levels, but to avoid changing the `ttrpc` crate at this point, I *could* change this PR to do something like this:

```diff
diff --git a/src/agent/Cargo.toml b/src/agent/Cargo.toml
index 28d19c04..7fcc83de 100644
--- a/src/agent/Cargo.toml
+++ b/src/agent/Cargo.toml
@@ -29,6 +29,7 @@ slog = { version = "2.5.2", features = ["dynamic-keys", "max_level_trace", "rele
 slog-scope = "4.1.2"

 # Redirect ttrpc log calls
+log = "0.4.11"
 slog-stdlog = "4.0.0"

 # for testing
diff --git a/src/agent/src/main.rs b/src/agent/src/main.rs
index 89fbf1ca..dae7f7da 100644
--- a/src/agent/src/main.rs
+++ b/src/agent/src/main.rs
@@ -67,6 +67,7 @@ mod test_utils;
 mod uevent;
 mod version;

+use log;
 use mount::{cgroups_mount, general_mount};
 use sandbox::Sandbox;
 use slog::Logger;
@@ -198,8 +199,12 @@ fn main() -> Result<()> {
     // which is required to satisfy the the lifetime constraints of the auto-generated gRPC code.
     let _guard = slog_scope::set_global_logger(logger.new(o!("subsystem" => "rpc")));

-    // Redirect ttrpc log calls to slog
-    let _log_guard = slog_stdlog::init()?;
+    let mut _log_guard: Result<(), log::SetLoggerError> = Ok(());
+
+    if config.log_level == slog::Level::Trace {
+        // Redirect ttrpc log calls to slog iff full debug requested
+        _log_guard = Ok(slog_stdlog::init().map_err(|e| e)?);
+    }

     start_sandbox(&logger, &config, init_mode)?;
```

**Comment by jodh-intel** [comment]
In other words, that will retain the existing logging behaviour. But if `agent.log=trace` or `KATA_AGENT_LOG_LEVEL=trace`  are set, the agent will also display `ttrpc` crate logs.

**Comment by jodh-intel** [comment]
Marking as `dnm` for now until we decide on the best approach...

**Review by liubin — APPROVED** [review]
It's very helpful, thanks @jodh-intel

**Review by lifupan — APPROVED** [review]

**Comment by jodh-intel** [comment]
Removing dnm label but I do wonder if we will need to modify ttrpc crate to make it slightly less "chatty" when this lands...

File: src/agent/src/config.rs [hunk 2]
```diff
@@ -21,6 +21,8 @@ const DEFAULT_HOTPLUG_TIMEOUT: time::Duration = time::Duration::from_secs(3);
 const DEFAULT_CONTAINER_PIPE_SIZE: i32 = 0;
 const VSOCK_ADDR: &str = "vsock://-1";
 const VSOCK_PORT: u16 = 1024;
+
+// Environment variable used for development and testing
```
Comment (egernst) [inline src/agent/src/config.rs]: nit: should this commit be squashed? (sorry; GitHub isn't good at providing commit reviews)

**Review by egernst — COMMENTED** [review]
If I understand correctly, when you enable agent debug logging with this PR, we'll get a lot of ttrpc logs which aren't particularly helpful?

But, you're suggesting you could enhance PR to enable ttrpc debug logs *iff* the debug level is set to trace? If so, this sounds like a good idea. If not, can you clarify what this looks like for the user for enabling logs, and what they get from that setting?

**Comment by lifupan** [comment]
> @lifupan - it's a slight abuse of logging levels, but to avoid changing the `ttrpc` crate at this point, I _could_ change this PR to do something like this:
> 
> ```diff
> diff --git a/src/agent/Cargo.toml b/src/agent/Cargo.toml
> index 28d19c04..7fcc83de 100644
> --- a/src/agent/Cargo.toml
> +++ b/src/agent/Cargo.toml
> @@ -29,6 +29,7 @@ slog = { version = "2.5.2", features = ["dynamic-keys", "max_level_trace", "rele
>  slog-scope = "4.1.2"
> 
>  # Redirect ttrpc log calls
> +log = "0.4.11"
>  slog-stdlog = "4.0.0"
> 
>  # for testing
> diff --git a/src/agent/src/main.rs b/src/agent/src/main.rs
> index 89fbf1ca..dae7f7da 100644
> --- a/src/agent/src/main.rs
> +++ b/src/agent/src/main.rs
> @@ -67,6 +67,7 @@ mod test_utils;
>  mod uevent;
>  mod version;
> 
> +use log;
>  use mount::{cgroups_mount, general_mount};
>  use sandbox::Sandbox;
>  use slog::Logger;
> @@ -198,8 +199,12 @@ fn main() -> Result<()> {
>      // which is required to satisfy the the lifetime constraints of the auto-generated gRPC code.
>      let _guard = slog_scope::set_global_logger(logger.new(o!("subsystem" => "rpc")));
> 
> -    // Redirect ttrpc log calls to slog
> -    let _log_guard = slog_stdlog::init()?;
> +    let mut _log_guard: Result<(), log::SetLoggerError> = Ok(());
> +
> +    if config.log_level == slog::Level::Trace {
> +        // Redirect ttrpc log calls to slog iff full debug requested
> +        _log_guard = Ok(slog_stdlog::init().map_err(|e| e)?);
> +    }
> 
>      start_sandbox(&logger, &config, init_mode)?;
> ```

Yeah，just as @egernst pointed, it's a good idea to log the ttrpc only "trace" level is set to avoid much more tedious ttrpc logs.

**Comment by jodh-intel** [comment]
@egernst - yes, that's exactly what I mean ;)

@lifupan, @egernst - I've added an extra commit to enable "only log ttrpc messages if log level is set to 'trace'".

**Comment by lifupan** [comment]
> /test

Hi @jodh-intel 
Some required CI  failed.

**Comment by Tim-Zhang** [comment]
@jodh-intel The file src/agent/Cargo.lock conflicts with target branch

**Comment by jodh-intel** [comment]
Thanks @Tim-Zhang - branch updated.

/test

**Pull request merged**